### ISSUE TITLE
Python/C++ API parity master tracking task

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#33227]
Achieving API parity between our Python and C++ frontend is crucial for broadening the number of language communities PyTorch can address. By the PyTorch 1.5 release, we will provide the following improvements to the PyTorch C++ frontend:

**Python/C++ API Parity:**
- [ ] torch.nn modules and functional (tracking issue: https://github.com/pytorch/pytorch/issues/25883). Items that will be added in 1.5:
  - [ ] RNN
  - [ ] LSTM
  - [ ] GRU
  - [ ] RNNCell
  - [ ] LSTMCell
  - [ ] GRUCell
- [ ] torch.optim optimizers (tracking issue: https://github.com/pytorch/pytorch/issues/28440)
- [ ] tensor multi-dim indexing API (PR: https://github.com/pytorch/pytorch/pull/32841 and others)
- [ ] C++ tensor autograd API (tracking issue: https://github.com/pytorch/pytorch/issues/25874). Remaining items:
  - [ ] grad_fn
  - [ ] register_hook 
  - [ ] retain_grad
  - [ ] _base

**Bug fixes:**
- [ ] Allow skipping default arguments in module's forward method when module is used in Sequential (PR: https://github.com/pytorch/pytorch/pull/33027)

**Docs:**
- [ ] For each `torch::nn` layer / functional, document how to use `*Options` to specify options
- [ ] Fix torch::Tensor doc generation (https://github.com/pytorch/pytorch/issues/25845)

**Tutorials:**
- [ ] How to use modern-style `torch::nn` layers and functionals (and deprecate use of `torch::nn::Functional`)
- [ ] How to use C++ tensor multi-dim indexing
- [ ] Autograd in C++
  - [ ] How to use C++ custom autograd function
  - [ ] How to use `at::Tensor::register_hook`
  - [ ] How to compute higher-order gradients in C++ (e.g. https://github.com/pytorch/pytorch/issues/18173)
